### PR TITLE
TLS-based RBAC and dynamic permission reconfiguration

### PR DESCRIPTION
**Pull request by kwart**
Demo showing dynamic permission reconfiguration. The authentication (and role assignment) is TLS-certificate-based.

The demo uses the OpenSSL Engine.

### CODE REVIEW

File: enterprise/tls-rbac-demo/00-create-keymaterial.sh [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+}
+
+prepareClientKeyMaterialExt() {
+  echo "Preparing client key material: $1, Extension: $2, Startdate: $3"
+  # Generate a key
+  openssl genpkey -outform PEM -out $1-key.pem -algorithm RSA -pkeyopt rsa_keygen_bits:2048
+  # Create a CSR request
+  openssl req -new -config ${OPENSSL_CONF} -keyform PEM -key $1-key.pem -outform PEM -out $1-cert.csr -subj "/C=US/O=Hazelcast/CN=$1"
+  # Issue a certificate
+  # openssl x509 -req -extensions v3_req -extfile ${OPENSSL_CONF} -inform DER -in $1-cert.csr -CA ca-cert.pem -CAkey ca-key.pem -CAcreateserial -outform DER -out $1-cert.crt
```
Comment (frant-hartm) [inline enterprise/tls-rbac-demo/00-create-keymaterial.sh:408]: why is this commented out?

File: enterprise/tls-rbac-demo/04-revoke-permissions.sh [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+java -cp target/tls-rbac-demo-*.jar com.hazelcast.samples.rbac.ReplacePermissions resources/no-permissions.xml
```
Comment (frant-hartm) [inline enterprise/tls-rbac-demo/04-revoke-permissions.sh:3]: No new line at end of file :-P

File: enterprise/tls-rbac-demo/README.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+* `TimestampClient`: runs a Hazelcast client that puts and gets timestamps into an `IMap`;
+* `ReplacePermissions`: dynamically changes the RBAC configuration by starting a new lite-member with a new permissions configuration. It gets a new permission configuration XML file path as the argument.
+
+The lite-member has the `on-join-operation` attribute configured to value `SEND` to distribute the new set of permissions.
+
+```java
+  config.getSecurityConfig().setOnJoinPermissionOperation(SEND);
+```
+
+[Look into the documentation](https://docs.hazelcast.com/imdg/4.1/security/native-client-security#handling-permissions-when-a-new-member-joins) for more details about handling permissions when a new member joins.
```
Comment (frant-hartm) [inline enterprise/tls-rbac-demo/README.md:16]: Should we use latest docs, instead IMDG 4.1 docs?

https://docs.hazelcast.com/hazelcast/latest/security/native-client-security#handling-permissions-when-a-new-member-joins

File: enterprise/tls-rbac-demo/resources/full-permissions.xml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<client-permissions>
+    <all-permissions principal="admin_user"/>
+
+    <map-permission name="timestamps" principal="regular_user">
+        <actions>
+            <action>all</action>
+        </actions>
+    </map-permission>
+</client-permissions>
```
Comment (frant-hartm) [inline enterprise/tls-rbac-demo/resources/full-permissions.xml:9]: new line again.

**Review by frant-hartm — CHANGES_REQUESTED** [review]

File: enterprise/tls-rbac-demo/04-revoke-permissions.sh (hunk 2 above)
Comment (kwart) [reply enterprise/tls-rbac-demo/04-revoke-permissions.sh:3]: :+1: will fix

File: enterprise/tls-rbac-demo/README.md (hunk 3 above)
Comment (kwart) [reply enterprise/tls-rbac-demo/README.md:16]: I don't like using the `latest` as the URL could change (refactoring, chapters reordering, etc.). I'll use the 5.0 URL instead.

**Review by frant-hartm — APPROVED** [review]

**Pull request merged**